keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+download and inspect the selected security profile before creating/upgrading a cluster,
+and setup command line flags, config files properly according to the rules during cluster creation/upgrade.
+
+These rules can also be interpreted as a set of check conditions for tools to validate if the cluster has been setup properly.
+
+Please see [Predefined Rules](#predefined-rules) as a reference of the rules.
+
+As currently, there's no way (or limited) to inspect the host operating system from inside the cluster,
+these rules are only enforced by the deployers (or cluster lifecycle management system).
+
+#### Runtime Rules
+Runtime rules are enforced after the cluster is bootstrapped,
+by _enforcers_ (the controllers deployed with security profiles in the cluster).
+Each rule is defined as an individual Custom Resource, that helps the enforcers define strong typed schemas.
+In a security profile, a rule is a simple reference to a custom resource.
+
+_Alternative design_
+
+Runtime rules can also be defined similarly to bootstrapping rules, in a schemaless way.
+However it loses the support from Kubernetes API to perform schema validation and other functionality,
+even though it's currently very limited for CRDs, in future it may become better.
+
+#### Effectiveness
+Multiple security profiles can exist in the same cluster at the same time,
+but only one can be active (being enforced with `runtime-rules`, not `bootstrapping-rules`).
+To specify which security profile is effective,
+the cluster admin creates a singleton cluster-scope custom resource of kind `SecurityProfileSelector`, with predefined name `current`:
+
+```yaml
+apiVersion: profile.security.k8s.io/v1alpha1
+kind: SecurityProfileSelector
+metadata:
+  name: current
+spec:
+  enforce: default-1.0-0
+```
+
+#### Switch Security Profiles
+
+Switching between security profiles is easy by updating `SecurityProfileSelector`.
+After the switch, some of the objects created by the previous security profile may no longer be desired.
+To take care of these objects properly,
+a cleanup mechanism is introduced with help of labeling the objects when they are created:
+
+```
+profile.security.k8s.io/profile-name=default-1.0-0
+profile.security.k8s.io/profile-rule=0
+```
+
+`profile-name` tracks the security profile,
+and `profile-rule` tracks the index in `runtime-rules` (see example above).
+After switch, the new enforcement controller is expected to behave:
+
+- Create/patch policy objects with updated labels;
+- Scanning and removing labeled objects of the same type but with mismatch values.
+
+#### Enforcement Status
+The controller/admission control enforcing a particular rule will be responsible to report the status of the enforcement.
+The status is defined as
+
+- `none`: no enforcers recognize this rule
+- `needed`: enforcement is needed, but not performed yet
+- `inprogress`: the enforcers are working to enforce the rule, not completed yet
```

Comment:
does 'inprogress' this only apply for runtime enforcers?

Reply:
yes